File: Arrays_Matrices/twoSum_method1_leetcode.py
# problem: https://leetcode.com/problems/two-sum/
# - TIME: O(n)
# - SPACE: O(n) 
# - Optimal Solution using hashtable

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        
        # a + b = c, where a=currNum and c=target, but we can rewrite it like so...
        # b = c - a
        
        seenNumbers = {}
        
        for i in range(len(nums)):                      # loop through array. do nums[i] - target = -numToFindInDictionary
            numToFind = target - nums[i]
            
            if (numToFind in seenNumbers and i != seenNumbers[numToFind]):
                    return [i, seenNumbers[numToFind]]
                
            seenNumbers[nums[i]] = i                    # if there is no solution found, store the current number in our hashmap.
        

File: Arrays_Matrices/twoSum_method1_leetcode.java
// problem: https://leetcode.com/problems/two-sum/
// - TIME: O(n)
// - SPACE: O(n)
//  - THE optimal solution utilizing a hashmap and 1 loops.

class Solution {
    
    public int[] twoSum(int[] nums, int target) {
        
        // populate the hashmap. <k,v> = <nums[idx], idx>
        HashMap<Integer, Integer> numsHM = new HashMap<Integer, Integer>();
        
        for(int i=0; i<nums.length; i++){
            int numToFind = target-nums[i];
            
            if(numsHM.containsKey(numToFind)){
                int ansArray[] = new int[2];
                ansArray[0] = i;
                ansArray[1] = numsHM.get(numToFind);
                return ansArray;
            }
            
            numsHM.put(nums[i], i);
        }
        
        return null; // should never hit here
    }
    
}